### ISSUE TITLE
windows MinGW64 build error

### ISSUE THREAD

**Issue by uliss** [Genivia/RE-flex#100]
Hello, found this error building in MinGW64 

```
make  all-recursive
make[1]: Entering directory '/c/Users/serge/work/RE-flex'
Making all in lib
make[2]: Entering directory '/c/Users/serge/work/RE-flex/lib'
g++ -DHAVE_CONFIG_H -I. -I..  -I../include -mavx2 -DHAVE_AVX2   -Wall -Wextra -Wunused -O2 -MT libreflex_a-matcher.o -MD -MP -MF .deps/libreflex_a-matcher.Tpo -c -o libreflex_a-matcher.o `test -f 'matcher.cpp' || echo './'`matcher.cpp
In file included from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
../include/reflex/absmatcher.h: In member function 'virtual void reflex::AbstractMatcher::reset(const char*)':
../include/reflex/absmatcher.h:352:13: error: '::posix_memalign' has not been declared
  352 |       if (::posix_memalign(reinterpret_cast<void**>(&buf_), 4096, max_) != 0)
      |             ^~~~~~~~~~~~~~
make[2]: *** [Makefile:529: libreflex_a-matcher.o] Error 1
make[2]: Leaving directory '/c/Users/serge/work/RE-flex/lib'
make[1]: *** [Makefile:456: all-recursive] Error 1
make[1]: Leaving directory '/c/Users/serge/work/RE-flex'
make: *** [Makefile:354: all] Error 2
```

**gcc version:**
```$ gcc --version
gcc.exe (Rev6, Built by MSYS2 project) 10.2.0
Copyright (C) 2020 Free Software Foundation, Inc.```

**Comment by genivia-inc** [comment]
Does `posix_memalign` without `::` work? The latest versions of MinGW64 support `posix_memalign` I have been informed.

**Comment by uliss** [comment]
the version I'm using is quit new:
```
gcc.exe (Rev6, Built by MSYS2 project) 10.2.0
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

without `::` does not work:
```
g++ -DHAVE_CONFIG_H -I. -I..  -I../include -mavx2 -DHAVE_AVX2   -Wall -Wextra -Wunused -O2 -MT libreflex_a-matcher.o -MD -MP -MF .deps/libreflex_a-matcher.Tpo -c -o libreflex_a-matcher.o `test -f 'matcher.cpp' || echo './'`matcher.cpp
In file included from ../include/reflex/matcher.h:40,
                 from matcher.cpp:37:
../include/reflex/absmatcher.h: In member function 'virtual void reflex::AbstractMatcher::reset(const char*)':
../include/reflex/absmatcher.h:352:11: error: 'posix_memalign' was not declared in this scope
  352 |       if (posix_memalign(reinterpret_cast<void**>(&buf_), 4096, max_) != 0)
      |           ^~~~~~~~~~~~~~
```

I am using standard MSYS2 under Windows 10, with Mingw64 tool-chain:
```
mingw-w64-x86_64-gcc 10.2.0-6
mingw-w64-x86_64-gcc-libgfortran 10.2.0-6
mingw-w64-x86_64-gcc-libs 10.2.0-6
mingw-w64-x86_64-libwinpthread-git
```

RE-flex versions <3.0 work and compile fine.

**Comment by uliss** [comment]
With this changes everything compiles

```c++
diff --git a/include/reflex/absmatcher.h b/include/reflex/absmatcher.h
index b7be179d..46dad98a 100644
--- a/include/reflex/absmatcher.h
+++ b/include/reflex/absmatcher.h
@@ -343,7 +343,7 @@ class AbstractMatcher {
     {
       max_ = 2 * Const::BLOCK;
 #if defined(WITH_REALLOC)
-#if (defined(__WIN32__) || defined(_WIN32) || defined(WIN32) || defined(_WIN64) || defined(__BORLANDC__)) && !defined(__CYGWIN__) && !defined(__MINGW32__) && !defined(__MINGW64__)
+#if (defined(__WIN32__) || defined(_WIN32) || defined(WIN32) || defined(_WIN64) || defined(__BORLANDC__)) && !defined(__CYGWIN__)
       buf_ = static_cast<char*>(_aligned_malloc(max_, 4096));
       if (buf_ == NULL)
         throw std::bad_alloc();
```

**Comment by genivia-inc** [comment]
Thank your for sharing your findings. The change makes sense.

The reason why older RE/flex versions do not have this issue is that aligned allocation of the larger lex buffer in the newer RE/flex version appears to improve performance.

I suspect that the aligned free should be changed accordingly, since aligned memory was allocated:
~~~
#if defined(WITH_REALLOC)
#if (defined(__WIN32__) || defined(_WIN32) || defined(WIN32) || defined(_WIN64) || defined(__BORLANDC__)) && !defined(__CYGWIN__)
      _aligned_free(static_cast<void*>(buf_));
~~~
While we are at it, also realloc should be changed to:
~~~
#if (defined(__WIN32__) || defined(_WIN32) || defined(WIN32) || defined(_WIN64) || defined(__BORLANDC__)) && !defined(__CYGWIN__)
        char *newbuf = static_cast<char*>(_aligned_realloc(static_cast<void*>(buf_), max_, 4096));
#else
~~~
Basically, all `!defined(__MINGW32__) && !defined(__MINGW64__)` conditions should be removed in this file. Please try that out and let me know if that works.

I just don't have the time to also start development and testing with MinGW and install MinGW. Everything is tested with Cygwin, Windows, MacOS, Linux Red Hat/Debian, Android.

**Comment by zachrammell** [comment]
I encountered the same build error and was able to fix it before finding this thread by removing the MinGW conditional you mentioned.
PR #101 fixes this issue.
